### PR TITLE
Lesson - Results

### PR DESCRIPTION
**Pull request by vinhnd2010**

### CODE REVIEW

File: app/assets/javascripts/timer.js [hunk 1]
```diff
@@ -0,0 +1,21 @@
+function time_remaining(total_times){
+  if(total_times <= 0){
```
Comment (ThuBM) [inline app/assets/javascripts/timer.js]: coding convention

File: app/assets/javascripts/timer.js [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+    $("#time-remaining").text(time_over_msg);
+    if(total_times == 0){
+      $("form").submit();
+    }
+    $("[type='submit']").remove();
+    return;
```
Comment (ThuBM) [inline app/assets/javascripts/timer.js]: remove

File: app/controllers/lessons_controller.rb [hunk 3]
```diff
@@ -1,25 +1,57 @@
...
+    @words = @category.words.not_learned(current_user).random_questions
+    @lesson = Lesson.new category_id: @category.id, user_id: current_user.id
     if @lesson.save
       flash[:success] = t "flash.lesson.created.success"
       redirect_to category_lesson_path @category, @lesson
     else
       flash[:danger] = t "flash.lesson.created.fails"
       redirect_to request.referrer
+      redirect_to category_lesson_path @category, @lesson
+    else
```
Comment (ThuBM) [inline app/controllers/lessons_controller.rb]: 2 else?????

File: app/models/lesson.rb [hunk 4]
```diff
@@ -2,27 +2,46 @@ class Lesson < ActiveRecord::Base
...
-  validates :name, presence: true
-  validates :description, presence: true
+  validate :check_words_left, on: :create
+
+  def num_of_correct_ans
+    self.results.select{|result| !result.answer.nil? && result.answer.correct}.length
+  end
+
+  def time_remaining
+    Settings.lesson.duration_in_minutes*60 - (Time.zone.now - self.created_at).to_i
```
Comment (ThuBM) [inline app/models/lesson.rb]: coding convention

File: config/routes.rb [hunk 5]
```diff
@@ -9,12 +9,9 @@
...
   resources :words, only: [:index]
-  resources :lessons, only: [:index, :create] do
-    resources :results
-  end
-
-  resources :categories, only: [:index] do
-    resources :lessons, only: [:index, :show, :create]
+  resources :lessons, only: [:index, :show]
+  resources :categories, only: [:index, :show] do
+    resources :lessons
```
Comment (ThuBM) [inline config/routes.rb]: only

File: app/controllers/lessons_controller.rb [hunk 6]
```diff
@@ -1,25 +1,57 @@
...
+      full_messages = ""
+      @lesson.errors.full_messages.each do |message|
```
Comment (ThuBM) [inline app/controllers/lessons_controller.rb]: sử dụng join

**Comment by ThuBM** [comment]
OK

File: app/controllers/categories_controller.rb [hunk 7]
```diff
@@ -2,6 +2,13 @@ class CategoriesController < ApplicationController
   before_action :logged_in_user
 
   def index
-    @categories = Category.paginate page: params[:page]
+    @categories = Category.order("name")
+      .paginate(page: params[:page], per_page: Settings.paginate_per_page)
```
Comment (dieunb) [inline app/controllers/categories_controller.rb]: remove `()`

**Pull request merged**